packages/timeline-state-resolver/src/integrations/tcpSend/index.ts

```diff
@@ -285,131 +92,98 @@ export class TCPSendDevice extends DeviceWithState<TSCSendState, DeviceOptionsTC
+		if (command.commandName === 'added' || command.commandName === 'changed') {
+			this.activeLayers.set(command.layer, this.getActiveLayersHash(command))
+		} else if (command.commandName === 'removed') {
+			this.activeLayers.delete(command.layer)
+		}
 
-		if (!this._tcpClient) {
-			this._tcpClient = new Socket()
-			this._tcpClient.on('connect', () => {
-				this._setConnected(true)
-			})
-			this._tcpClient.on('close', () => {
-				this._setConnected(false)
-			})
-			this._tcpClient.on('end', () => {
-				this._setConnected(false)
-			})
+		if (command.layer && command.commandName !== 'manual') {
+			const hash = this.activeLayers.get(command.layer)
+			if (this.getActiveLayersHash(command) !== hash) return Promise.resolve() // command is no longer relevant to state
 		}
-		if (!this.connected) {
-			return new Promise((resolve, reject) => {
-				this._tcpClient!.connect(this._port, this._host, () => {
-					resolve()
-					// client.write('Hello, server! Love, Client.');
-				})
-				setTimeout(() => {
-					reject(`TCP timeout: Unable to connect to ${this._host}:${this._port}`)
-				}, TIMEOUT)
-			})
-		} else {
+		if (this._terminated) {
 			return Promise.resolve()
 		}
-	}
-	private async _sendTCPMessage(message: string): Promise<void> {
-		// Do we have a client?
-		return this._connectTCPClient().then(() => {
-			if (this._tcpClient) {
-				this._tcpClient.write(Buffer.from(message, this._bufferEncoding))
-			} else throw Error('_sendTCPMessage: _tcpClient is falsy!')
-		})
-	}
-	private async _defaultCommandReceiver(
-		_time: number,
-		cmd: TcpSendCommandContent,
-		context: CommandContext,
-		timelineObjId: string
-	): Promise<any> {
-		// this.emit('info', 'TCTSend ', cmd)
 
-		const cwc: CommandWithContext = {
-			context,
-			command: cmd,
-			timelineObjId,
-		}
-		this.emitDebug(cwc)
+		this.context.emitDebug({ context, timelineObjId, command })
 
-		if (cmd.message) {
-			return this._sendTCPMessage(cmd.message)
-		} else {
-			return Promise.reject('tcpCommand.message not set')
+		await this.tcpConnection.sendTCPMessage(command.content.message)
+	}
+	private async actionSendTcpCommand(cmd?: TcpSendCommandContent): Promise<ActionExecutionResult> {
+		if (!cmd)
+			return {
+				result: ActionExecutionResultCode.Error,
+				response: t('Failed to send command: Missing payload'),
+			}
+		if (!cmd.message) {
+			return {
+				result: ActionExecutionResultCode.Error,
+				response: t('Failed to send command: Missing message'),
+			}
+		}
+		try {
+			await this.sendCommand({
+				timelineObjId: '',
+				context: 'makeReady',
+				command: {
+					commandName: 'manual',
+					content: cmd,
+					layer: '',
+				},
+			})
+		} catch (e) {
+			this.context.emitWarning('Manual TCP command failed: ' + JSON.stringify(cmd))
+			return {
+				result: ActionExecutionResultCode.Error,
+				response: t('Error when sending TCP command: {{errorMessage}}', { errorMessage: `${error}` }),
```

Comment:
```suggestion
				response: t('Error when sending TCP command: {{errorMessage}}', { errorMessage: `${e}` }),
```

This is not using the correct variable